{% extends "layout.html" %}

{% block content %}
<br>
<div class="container">
    <div class="row">
        <div class="col-md-12 text-center">
            <h1>Welcome to Immunolyser 1.0</h1>
            <br>
            <div class="text-justify">
                <p>Immunolyser is an automated web-based tool developed for immunologists to analyze immunopeptidomics data in a seamless pipeline. It integrates multiple computational tools for peptide analysis, generating a comprehensive report summarizing sequence conservation, clustering of similar peptides, and predicting binding affinity to various MHC allotypes based on uploaded peptide data. The user-friendly interface requires no programming experience.</p>
                
                <p>For detailed guidelines on using the pipeline and analyzing the web-based output reports, please refer to the <a href="{{ url_for('help') }}">help</a> page.</p>

                <p>If you find Immunolyser helpful, please consider citing us in your study:<br>
                    <i>Prithvi Raj Munday, Joshua Fehring, Jerico Revote, Kirti Pandey, Mohammad Shahbazy, Katherine E Scull, Sri H. Ramarathinam, Pouya Faridi, Nathan P. Croft, Asolina Braun, Chen Li, and Anthony W. Purcell. Immunolyser: a web-based computational pipeline for analyzing and mining immunopeptidomic data. Computational and Structural Biotechnology Journal, 2023.</i>
                </p>
                
                <h4>How to use Immunolyser?</h4>
                <p>Use the <a href="{{ url_for('initialiser') }}">Initialiser</a> module to submit a job. The following inputs are required:</p>
                <ul>
                    <li><b>Sample file(s)</b>: A CSV file with a 'Peptide' column. Additional columns like 'Length' and 'Accession' are optional. Multiple samples and replicate files can be uploaded.</li>
                    <li><b>Control file(s)</b> (optional): A CSV file with a 'Peptide' column. Peptides in this file are labeled as 'control' in results. Multiple control files can be uploaded.</li>
                    <li><b>Alleles of interest</b> (optional): MHC-I/II alleles for peptide-MHC binding predictions.</li>
                    <li><font color="red"><b>Running time</b></font>: Please wait patiently as Immunolyser processes the data. Do not refresh or close the page.</li>
                </ul>
                
                <h4>Demo and Example Input</h4>
                <p>A demo report is available to demonstrate the analysis results. Access it via the <a href="{{ url_for('demo') }}">Demo</a> tab. To test the pipeline, download example files from <a href="/static/temp/Study by Pandey et al Dataset.zip" target="_blank">here</a> and upload them in the <a href="{{ url_for('initialiser') }}">initialiser</a> module. Additional example input files from other search engines can be downloaded <a href="/static/temp/demo data from multiple softwares.zip" target="_blank">here</a>.</p>
                
                <h4>References</h4>
                <p>The following tools are used in the pipeline:</p>
                <ol>
                    <li><b>Seq2Logo 2.0</b>: Construction and visualization of amino acid binding motifs and sequence profiles. <i>Martin Christen Frolund Thomsen; Morten Nielsen, Nucleic Acids Research 2012; 40 (W1): W281-W287</i>.</li>
                    <li><b>GibbsCluster 2.0</b>: Unsupervised clustering and alignment of peptide sequences. <i>Andreatta M, Alvarez B, Nielsen M, Nucleic Acids Research (2017) doi: 10.1093/nar/gkx248</i>.</li>
                    <li><b>NetMHCpan 4.1</b>: Improved predictions of MHC antigen presentation. <i>Birkir Reynisson, Bruno Alvarez, Sinu Paul, Bjoern Peters, and Morten Nielsen, Nucleic Acids Research, May 2020, https://doi.org/10.1093/nar/gkaa379</i>.</li>
                    <li><b>Anthem</b>: Tool for fast and accurate prediction of binding between peptides and HLA class I molecules. <i>Mei S, Li F, Xiang D, et al. Brief Bioinform 2021. doi:10.1093/bib/bbaa415</i>.</li>
                    <li><b>MixMHCpred 2.1</b>: <i>Bassani-Sternberg M et al. Deciphering HLA motifs across HLA peptidomes, PLoS Comp Bio (2017)</i> and <i>Gfeller et al. The length distribution and multiple specificity of naturally presented HLA-I ligands, J Immunol (2018)</i>.</li>
                    <li><b>MixMHC2pred</b>: <i>Racle J, Michaux J, Rockinger GA et al. Robust prediction of HLA class II epitopes, Nat Biotechnol 2019;37:1283-1286</i>.</li>
                </ol>
                
                <h4>Feedback</h4>
                <p>Your feedback helps us improve. Please share your feedback at <a href="mailto:Chen.Li@monash.edu">Chen.Li@monash.edu</a>.</p>
            </div>
        </div>
    </div>
</div>
<br>
{% endblock %}
